applications/GeoMechanicsApplication/custom_elements/triangle_2d_15.hpp

```diff
@@ -0,0 +1,1752 @@
+                shape_function_values(pnt, 1)  = xi * (xi - 0.25) * (xi - 0.50) * (xi - 0.75) * cof2;
+                shape_function_values(pnt, 2)  = et * (et - 0.25) * (et - 0.50) * (et - 0.75) * cof2;
+                shape_function_values(pnt, 3)  = xi * zt * (zt - 0.25) * (zt - 0.50) * cof1;
+                shape_function_values(pnt, 4)  = xi * zt * (zt - 0.25) * (xi - 0.25) * 64.0;
+                shape_function_values(pnt, 5)  = xi * zt * (xi - 0.25) * (xi - 0.50) * cof1;
+                shape_function_values(pnt, 6)  = xi * et * (xi - 0.25) * (xi - 0.50) * cof1;
+                shape_function_values(pnt, 7)  = xi * et * (xi - 0.25) * (et - 0.25) * 64.0;
+                shape_function_values(pnt, 8)  = xi * et * (et - 0.25) * (et - 0.50) * cof1;
+                shape_function_values(pnt, 9)  = et * zt * (et - 0.25) * (et - 0.50) * cof1;
+                shape_function_values(pnt, 10) = et * zt * (et - 0.25) * (zt - 0.25) * 64.0;
+                shape_function_values(pnt, 11) = et * zt * (zt - 0.25) * (zt - 0.50) * cof1;
+                shape_function_values(pnt, 12) = xi * et * zt * (zt - 0.25) * 128.0;
+                shape_function_values(pnt, 13) = xi * et * zt * (xi - 0.25) * 128.0;
+                shape_function_values(pnt, 14) = xi * et * zt * (et - 0.25) * 128.0;
+            }
+            return shape_function_values;
+        }
+
+        // ========================================================================================
+        // ========================================================================================
+        /**
+         * Calculates the local gradients of all shape functions
+         * in all integration points.
+         * Integration points are expected to be given in local coordinates
+         *
+         * @param ThisMethod the current integration method
+         * @return the vector of the gradients of all shape functions
+         * in each integration point
+         */
+        static ShapeFunctionsGradientsType
+            CalculateShapeFunctionsIntegrationPointsLocalGradients(
+                typename BaseType::IntegrationMethod ThisMethod)
+        {
+            IntegrationPointsContainerType all_integration_points = AllIntegrationPoints();
+            IntegrationPointsArrayType integration_points = all_integration_points[static_cast<int>(ThisMethod)];
+            //number of integration points
+            const int integration_points_number = integration_points.size();
+            ShapeFunctionsGradientsType d_shape_f_values(integration_points_number);
+            //initialising container
+            //std::fill(d_shape_f_values.begin(), d_shape_f_values.end(), Matrix(4,2));
+            //
+            //loop over all integration points
+            for (int pnt = 0; pnt < integration_points_number; pnt++)
+            {
+                Matrix result(15, 2);
+                double xi1 = integration_points[pnt].X();
+                double xi2 = xi1 * xi1;
```

Comment:
const?

Reply:
These variables are in for loop and they are overwritten for each integration point.

Reply:
Set const in the loop, works fine. Done